src/lib/http-observer.js

```diff
@@ -0,0 +1,220 @@
+import Uri from 'urijs';
+
+import Logger from './log';
+import { CDN_HEADERS, CDN_HOSTNAMES } from './cdn';
+
+const logger = new Logger('webext.lib.http-observer', console);
+
+export const REQUEST_TYPES = [
+  'main_frame',
+  'sub_frame',
+  'stylesheet',
+  'script',
+  'image',
+  'object',
+  'xmlhttprequest',
+  'xbl',
+  'xslt',
+  'ping',
+  'beacon',
+  'xml_dtd',
+  'font',
+  'media',
+  'websocket',
+  'csp_report',
+  'imageset',
+  'web_manifest',
+  'other'
+];
+
+export default class HttpObserver {
+  constructor() {
+    this.requests = {};
+    this.addListeners();
+  }
+
+  addListeners() {
+    browser.webNavigation.onBeforeNavigate.addListener(details => {
+      this.navigationListener(details);
+    });
+    browser.webRequest.onBeforeRequest.addListener(
+      details => this.requestListener(details),
+      { urls: [ '<all_urls>' ] }
+    );
+    browser.webRequest.onCompleted.addListener(
+      details => this.responseListener(details),
+      { urls: [ '<all_urls>' ] },
+      [ 'responseHeaders' ]
+    );
+  }
+
+  // Whenever the user navigates to a new page, clear the collected requests
+  // for this tab.
+  navigationListener(details) {
+    this.requests[details.tabId] = {};
+  }
+
+  // Whenever a request is made, log some initial metadata. Only do this if the
+  // instance was listening at webNavigation.onBeforeNavigate.
+  requestListener(details) {
+    if (details.tabId in this.requests) {
+      this.requests[details.tabId][details.requestId] = {
+        start: details.timeStamp,
+        url: details.url,
+        completed: false
+      };
+    }
+  }
+
+  // Whenever a request is completed, log additional metadata. Only do this if
+  // the instance was listening at webNavigation.onBeforeNavigate.
+  responseListener(details) {
+    if (details.tabId in this.requests) {
+      const request = this.requests[details.tabId][details.requestId];
```

Comment:
Seeing a `request is undefined` error occasionally here. I dropped a log statement in here like:
```
console.log('responseListener', details.tabId, details.requestId, Object.keys(this.requests[details.tabId])); // eslint-disable-line no-console
```
Trying to dig into the cause, but FWIW this is the tail end of console output I'm seeing when I visit qz.com:
<img width="1123" alt="screenshot 2017-02-15 17 04 31" src="https://cloud.githubusercontent.com/assets/21687/22997569/ea53e3a2-f3a0-11e6-9d35-095760d8704c.png">